src/components/UIInventory.cpp

```diff
@@ -0,0 +1,227 @@
+      //        I don't know how yet.
+      mViewedInventory->OnItemChanged.clear();
+
+      REGOTH_LOG(Warning, Uncategorized, "[UIInventory] Cleared Inventory OnItemChanged callback!");
+    }
+
+    mViewedInventory = inventory;
+
+    if (mRegisteredOnItemChangedEvent)
+    {
+      mRegisteredOnItemChangedEvent.disconnect();
+    }
+
+    if (mViewedInventory)
+    {
+      mRegisteredOnItemChangedEvent =
+          mViewedInventory->OnItemChanged.connect([this](const bs::String& instance) {
+            // Get notified when the viewed inventory changes
+            onInventoryItemUpdated(instance);
+          });
+
+      forceUpdateAll();
+    }
+  }
+
+  void UIInventory::onInventoryItemUpdated(const bs::String& instance)
+  {
+    bs::UINT32 countUI        = itemCount(instance);
+    bs::UINT32 countInventory = mViewedInventory->itemCount(instance);
+
+    if (countUI < countInventory)
+    {
+      addItem(instance, instance, countInventory - countUI);
+    }
+    else if (countUI > countInventory)
+    {
+      removeItem(instance, countUI - countInventory);
+    }
+    else
+    {
+      // Count is the same, do nothing
+    }
+  }
+
+  void UIInventory::update()
+  {
+    UIElement::update();
+
+    bs::Rect2I bounds = parentLayout().getBounds();
+
+    bounds.y      = 0;
+    bounds.height = parentLayout().getBounds().height;
+    bounds.width  = 180;
+
+    layout().setBounds(bounds);
+    mScrollArea->setWidth(layout().getBounds().width);
+    mScrollArea->setHeight(layout().getBounds().height);
+  }
+
+  void UIInventory::removeAll()
+  {
+    for (const auto& instance : allInstancesInInventory())
+    {
+      removeUIElementsFromItem(mItemsByInstance[instance]);
+    }
+
+    mItemsByInstance.clear();
+  }
+
+  void UIInventory::forceUpdateAll()
+  {
+    removeAll();
+
+    for (const auto& p : mViewedInventory->allItems())
+    {
+      onInventoryItemUpdated(p.first);
+    }
+  }
+
+  bs::Vector<bs::String> UIInventory::allInstancesInInventory() const
+  {
+    bs::Vector<bs::String> instances;
+
+    instances.reserve(mItemsByInstance.size());
+
+    for (const auto& p : mItemsByInstance)
+    {
+      instances.push_back(p.first);
+    }
+
+    return instances;
+  }
+
+  bool UIInventory::hasItem(const bs::String& instance) const
+  {
+    return itemCount(instance) != 0;
+  }
+
+  bs::UINT32 UIInventory::itemCount(const bs::String& instance) const
+  {
+    auto it = mItemsByInstance.find(instance);
+
+    if (it == mItemsByInstance.end()) return 0;
+
+    return it->second.count;
+  }
+
+  void UIInventory::addItem(const bs::String& instance, const bs::String& name, bs::UINT32 count)
+  {
+    auto it = mItemsByInstance.find(instance);
+
+    if (it == mItemsByInstance.end())
+    {
+      auto& item = mItemsByInstance[instance];
+
+      item.count = count;
+      item.name  = name;
+
+      createUIElementsForItem(item);
+
+      updateItemThumbnail(item);
```

Comment:
Is there a usecase where you would create an UIElement for the item but not update its Thumbnail? 
If not maybe we should move it into the creation of the ui element.

Reply:
I kept those separated because then I feel like I'd have to call that method `createUIElementsForItemAndFillThem` or something. `createUIElementsForItem` is supposed to *only create* it, not access any script data, create textures, etc.